glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,325 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+use std::fs;
+use std::fs::{create_dir_all, File, OpenOptions};
+use std::io::BufRead;
+use std::io::BufReader;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::sync::RwLock;
+
+use serde::{Deserialize, Serialize};
+use serde_json;
+use serde_json::{json, Value as JsonValue};
+
+use crate::CommonMetricData;
+use crate::Glean;
+use crate::Result;
+
+/// Represents the data for a single event.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct RecordedEventData {
+    pub timestamp: u64,
+    pub category: String,
+    pub name: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub extra: Option<HashMap<String, String>>,
+}
+
+impl RecordedEventData {
+    /// Serialize an event to JSON, adjusting its timestamp relative to a base timestamp
+    pub fn serialize_relative(&self, timestamp_offset: u64) -> JsonValue {
+        json!(&RecordedEventData {
+            timestamp: self.timestamp - timestamp_offset,
+            category: self.category.clone(),
+            name: self.name.clone(),
+            extra: self.extra.clone(),
+        })
+    }
+}
+
+/// This struct handles the in-memory and on-disk storage logic for events.
+///
+/// So that the data survives shutting down of the application, events are stored
+/// in an append-only file on disk, in addition to the store in memory. Each line
+/// of this file records a single event in JSON, exactly as it will be sent in the
+/// ping. There is one file per store.
+///
+/// When restarting the application, these on-disk files are checked, and if any are
+/// found, they are loaded, queued for sending and flushed immediately before any
+/// further events are collected. This is because the timestamps for these events
+/// may have come from a previous boot of the device, and therefore will not be
+/// compatible with any newly-collected events.
+#[derive(Debug)]
+pub struct EventDatabase {
+    pub path: PathBuf,
+    // The in-memory list of events
+    event_stores: RwLock<HashMap<String, Vec<RecordedEventData>>>,
+    file_lock: RwLock<()>,
+}
+
+impl EventDatabase {
+    /// Create a new event database.
+    ///
+    /// # Arguments
+    ///
+    /// * `data_path` - The directory to store events in. A new directory
+    ///   `events` will be created inside of this directory.
+    pub fn new(data_path: &str) -> Result<Self> {
+        let path = Path::new(data_path).join("events");
+        create_dir_all(&path)?;
+
+        Ok(Self {
+            path,
+            event_stores: RwLock::new(HashMap::new()),
+            file_lock: RwLock::new(()),
+        })
+    }
+
+    /// Initialize events storage. This must be called once on application startup,
```

Comment:
The first sentence here doesn't reflect much of the actual function name. Maybe we can clarify this further?

"Initialize events storage after Glean is fully initialized and ready to send pings" or something along the lines.
(Just making it clear in the first sentence what it does and when it gets called)